mono/utils/mono-windows-thread-name.c

```diff
@@ -0,0 +1,149 @@
+// mono-windows-thread-name.c
+//
+// There several ways to set thread name on Windows.
+//
+// Historically: Raise an exception with an ASCII string.
+//  This is visible only in a debugger. Only if a debugger
+//  is running when the exception is raised. There is
+//  no way to get the thread name and they do not appear in ETW traces.
+//
+// XboxOne: SetThreadName(thread handle, unicode)
+//   This works with or without a debugger and can be retrieved with GetThreadName.
+//   Sometimes an inline function SetThreadDescription for source compat with next.
+//   https://github.com/microsoft/xbox-live-api/blob/90b38b434d9c13ce4916c116cd28a98b239e38e2/InProgressSamples/Kits/ATGTK/ThreadHelpers.h#L21
+//
+// Windows 10 1607 or newer (according to documentation, or Creators Update says https://randomascii.wordpress.com/2015/10/26/thread-naming-in-windows-time-for-something-better).
+//  SetThreadDescription(thread handle, unicode)
+//  This is like XboxOne -- works with or without debugger, can be retrieved
+//  with GetThreadDescription, and appears in ETW traces.
+//  See https://randomascii.wordpress.com/2015/10/26/thread-naming-in-windows-time-for-something-better.
+//
+//  This is not called SetThreadName to avoid breaking compilation of C (but not C++)
+//  copied from http://msdn.microsoft.com/en-us/library/xcb2z8hs.aspx.
+//
+// UWP: Also SetThreadDescription, but running prior to 1607 (or Creators Update per above).
+//   would require LoadLibrary / GetProcAddress, but LoadLibrary is not allowed.
+//
+// Author:
+//  Jay Krell (jaykrell@microsoft.com)
+//
+// Copyright 2019 Microsoft
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+#include "config.h"
+#include "mono-threads.h"
```

Comment:
`#include` should be move to after `#if`, but it doesn't matter, the file is also filtered in existing Makefiles.